src/backend/bpf.rs

```diff
@@ -0,0 +1,213 @@
+//! Define constants, helpers and types used for BPF codegen.
+
+use super::TargetArch;
+
+/// Program made up of a sequence of BPF instructions.
+pub type BpfProgram = Vec<sock_filter>;
+
+/// Reference to program made up of a sequence of BPF instructions.
+pub type BpfProgramRef<'a> = &'a [sock_filter];
+
+// The maximum number of a syscall argument.
+// A syscall can have at most 6 arguments.
+// Arguments are numbered from 0 to 5.
+pub const ARG_NUMBER_MAX: u8 = 5;
+
+// The maximum number of BPF statements that a condition will be translated into.
+// This is not a linux requirement but is based on the current BPF compilation logic.
+// This is used in the backend code for preallocating vectors and for detecting unjumpable offsets.
+pub const CONDITION_MAX_LEN: u8 = 6;
+
+// The maximum seccomp-BPF program length allowed by the linux kernel.
+pub const BPF_MAX_LEN: usize = 4096;
+
+// `struct seccomp_data` offsets and sizes of fields in bytes:
+//
+// ```c
+// struct seccomp_data {
+//     int nr;
+//     __u32 arch;
+//     __u64 instruction_pointer;
+//     __u64 args[6];
+// };
+// ```
+pub const SECCOMP_DATA_NR_OFFSET: u8 = 0;
+const SECCOMP_DATA_ARCH_OFFSET: u8 = 4;
+pub const SECCOMP_DATA_ARGS_OFFSET: u8 = 16;
+pub const SECCOMP_DATA_ARG_SIZE: u8 = 8;
+
+// Builds a `jump` BPF instruction.
+//
+// # Arguments
+//
+// * `code` - The operation code.
+// * `jt` - The jump offset in case the operation returns `true`.
+// * `jf` - The jump offset in case the operation returns `false`.
+// * `k` - The operand.
+#[allow(non_snake_case)]
+#[inline(always)]
+pub(crate) fn BPF_JUMP(code: u16, k: u32, jt: u8, jf: u8) -> sock_filter {
+    sock_filter { code, jt, jf, k }
+}
+
+// Builds a "statement" BPF instruction.
+//
+// # Arguments
+//
+// * `code` - The operation code.
+// * `k` - The operand.
+#[allow(non_snake_case)]
+#[inline(always)]
+pub(crate) fn BPF_STMT(code: u16, k: u32) -> sock_filter {
+    sock_filter {
+        code,
+        jt: 0,
+        jf: 0,
+        k,
+    }
+}
+
+// Builds a sequence of BPF instructions that validate the underlying architecture.
+#[allow(non_snake_case)]
+#[inline(always)]
+pub(crate) fn VALIDATE_ARCHITECTURE(target_arch: TargetArch) -> Vec<sock_filter> {
```

Comment:
nit: the name is misleading. It sounds like something gets validated, but the function actually builds a architecture validator. Can you rename it?